identity/v3userpass.go

```diff
@@ -0,0 +1,182 @@
+}
+
+// v3AuthToken contains the token to use for token authentication.
+type v3AuthToken struct {
+	ID string `json:"id"`
+}
+
+// v3AuthScope contains the scope of the authentication request.
+type v3AuthScope struct {
+	Domain  *v3AuthDomain  `json:"domain,omitempty"`
+	Project *v3AuthProject `json:"project,omitempty"`
+}
+
+// v3AuthProject contains the project scope for the authentication
+// request.
+type v3AuthProject struct {
+	Domain *v3AuthDomain `json:"domain,omitempty"`
+	ID     string        `json:"id,omitempty"`
+	Name   string        `json:"name,omitempty"`
+}
+
+/// / V3UserPass is an Authenticator that will perform username + password
+// authentication using the v3 protocol.
+type V3UserPass struct {
+	client *goosehttp.Client
+}
+
+// Auth performs a v3 username + password authentication request using
+// the values supplied in creds.
+func (u *V3UserPass) Auth(creds *Credentials) (*AuthDetails, error) {
+	if u.client == nil {
+		u.client = goosehttp.New()
+	}
+	auth := v3AuthWrapper{
+		Auth: v3AuthRequest{
+			Identity: v3AuthIdentity{
+				Methods: []string{"password"},
+				Password: &v3AuthPassword{
+					User: v3AuthUser{
+						Domain: &v3AuthDomain{
+							ID: "default",
+						},
+						Name:     creds.User,
+						Password: creds.Secrets,
+					},
+				},
+			},
+		},
+	}
+	if creds.TenantName != "" {
+		auth.Auth.Scope = &v3AuthScope{
+			Project: &v3AuthProject{
+				Domain: &v3AuthDomain{
+					ID: "default",
+				},
+				Name: creds.TenantName,
+			},
+		}
+	}
+	return v3KeystoneAuth(u.client, &auth, creds.URL)
+}
+
+type v3TokenWrapper struct {
+	Token v3Token `json:"token"`
+}
+
+type v3Token struct {
+	Expires time.Time        `json:"expires_at"`
+	Issued  time.Time        `json:"issued_at"`
+	Methods []string         `json:"methods"`
+	Catalog []v3TokenCatalog `json:"catalog"`
+	Project v3TokenProject   `json:"project"`
+	User    v3TokenUser      `json:"user"`
+}
+
+type v3TokenCatalog struct {
+	ID        string            `json:"id"`
+	Type      string            `json:"type"`
+	Endpoints []v3TokenEndpoint `json:"endpoints"`
+}
+
+type v3TokenEndpoint struct {
+	ID        string `json:"id"`
+	Region    string `json:"region"`
+	URL       string `json:"url"`
+	Interface string `json:"interface"`
+}
+
+type v3TokenProject struct {
+	ID string `json:"id"`
+}
+
+type v3TokenUser struct {
+	ID string `json:"id"`
+}
+
+// v3KeystoneAuth performs a v3 authentication request.
+func v3KeystoneAuth(c *goosehttp.Client, v interface{}, url string) (*AuthDetails, error) {
+	var resp v3TokenWrapper
+	req := goosehttp.RequestData{
+		ReqValue:  v,
+		RespValue: &resp,
+		ExpectedStatus: []int{
+			http.StatusCreated,
+		},
+	}
+	if err := c.JsonRequest("POST", url, "", &req, nil); err != nil {
+		return nil, err
+	}
+	tok := req.RespHeaders.Get("X-Subject-Token")
+	if tok == "" {
+		return nil, fmt.Errorf("authentication failed")
+	}
+	rsu := make(map[string]ServiceURLs, len(resp.Token.Catalog))
```

Comment:
If I'm reading the API docs correctly, this code only caters for the case when "catalog" is returned inside a "token", but not when there is only "token", without "catalog" inside it.